nemo_text_processing/text_normalization/vi/verbalizers/time.py

```diff
@@ -0,0 +1,149 @@
+
+        def extract_field(field_name):
+            return (
+                pynutil.delete(f"{field_name}:")
+                + delete_space
+                + pynutil.delete("\"")
+                + quoted_text
+                + pynutil.delete("\"")
+            )
+
+        hour_component = extract_field("hours")
+        timezone_component = extract_field("zone") @ time_zone
+
+        zero_minute_component = pynutil.delete("minutes:") + delete_space + pynutil.delete("\"không\"")
+        zero_second_component = pynutil.delete("seconds:") + delete_space + pynutil.delete("\"không\"")
+
+        non_zero_minute_component = (
+            pynutil.delete("minutes:")
+            + delete_space
+            + pynutil.delete("\"")
+            + pynini.closure(NEMO_NOT_QUOTE - pynini.accep("không"))
+            + pynutil.delete("\"")
+        )
+        non_zero_second_component = (
+            pynutil.delete("seconds:")
+            + delete_space
+            + pynutil.delete("\"")
+            + pynini.closure(NEMO_NOT_QUOTE - pynini.accep("không"))
+            + pynutil.delete("\"")
+        )
+
+        hour_with_unit = hour_component + pynutil.insert(" giờ")
+        minute_with_unit = non_zero_minute_component + pynutil.insert(" phút")
+        second_with_unit = non_zero_second_component + pynutil.insert(" giây")
+
+        optional_timezone = pynini.closure(delete_space + pynutil.insert(" ") + timezone_component, 0, 1)
+        optional_preserve_order = pynini.closure(delete_space + delete_preserve_order, 0, 1)
+
+        patterns = [
+            hour_with_unit
+            + pynini.closure(delete_space + zero_minute_component, 0, 1)
+            + pynini.closure(delete_space + zero_second_component, 0, 1)
+            + optional_timezone
+            + optional_preserve_order,
+            hour_with_unit
+            + delete_space
+            + pynutil.insert(" ")
+            + minute_with_unit
+            + pynini.closure(delete_space + zero_second_component, 0, 1)
+            + optional_timezone
+            + optional_preserve_order,
+            hour_with_unit
+            + delete_space
+            + zero_minute_component
+            + delete_space
+            + pynutil.insert(" ")
+            + second_with_unit
+            + optional_timezone
+            + optional_preserve_order,
+            hour_with_unit
+            + delete_space
+            + pynutil.insert(" ")
+            + minute_with_unit
+            + delete_space
+            + pynutil.insert(" ")
+            + second_with_unit
+            + optional_timezone
+            + optional_preserve_order,
+            minute_with_unit + pynini.closure(delete_space + zero_second_component, 0, 1),
+            minute_with_unit + delete_space + pynutil.insert(" ") + second_with_unit,
+            second_with_unit,
+        ]
+
+        final_graph = pynini.union(*patterns)
+
+        if not deterministic:
```

Comment:
Any phrases equivalent to "quarter til XYZ" "half past ABC"?

Reply:
alright, may come up as a new pr in the future or you can tackle now. either or is fine.